File: src/TelecommsSimulation.Core.CS/BaseStation.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

using TelecommsSimulation.Utilities;

namespace TelecommsSimulation.Core;

public class BaseStation : ILocatable, INotifyPropertyChanged
{
    protected readonly ConcurrentDictionary<string, Phone> ConnectedPhonesMap = [];

    // Workaround: C++/CLI does not support System.Collections.Immutable even from C# NuGet reference (because it does not support NuGet altogether)
    // Hack: Also should be private, but it's public for the sake of connection visualization
    public IEnumerable<Phone> ConnectedPhones => ConnectedPhonesMap.Values.ToImmutableArray();

    protected static readonly ConcurrentDictionary<string, CallPhoneData> CallDataLookupDictionary = [];

    // Workaround: C++/CLI does not support System.Collections.Immutable even from C# NuGet reference (because it does not support NuGet altogether)
    // Hack: Also should not exist at all, but exists for the sake of calls visualization
    public static IEnumerable<CallPhoneData> CurrentCalls => CallDataLookupDictionary.Values.ToImmutableArray().Distinct();

    public string Name { get; }

    private bool _isEnabled;

    public bool IsEnabled
    {
        get => _isEnabled;
        private set => SetField(ref _isEnabled, value);
    }

    public Coordinates Location { get; }

    public double CoverageRadiusKm { get; }

    public int ConnectedPhonesCount => ConnectedPhonesMap.Count;

    protected static readonly List<BaseStation> AllEnabledStationsList = [];

    // Workaround: C++/CLI does not support System.Collections.Immutable even from C# NuGet reference (because it does not support NuGet altogether)
    public static IEnumerable<BaseStation> AllEnabledStations => AllEnabledStationsList.ToImmutableArray();

    protected readonly SynchronizationContext? UiSynchronizationContext;

    public BaseStation(string name, Coordinates location, double coverageRadiusKm)
    {
        UiSynchronizationContext = SynchronizationContext.Current;

        Name = name;
        Location = location;
        CoverageRadiusKm = coverageRadiusKm;
    }

    public virtual void Enable()
    {
        if (IsEnabled)
            throw new InvalidBaseStationStateException();

        IsEnabled = true;
        AllEnabledStationsList.Add(this);
    }

    public virtual void Disable()
    {
        if (!IsEnabled)
            throw new InvalidBaseStationStateException();

        AllEnabledStationsList.Remove(this);

        foreach (KeyValuePair<string, Phone> connectedPhonePair in ConnectedPhonesMap)
            connectedPhonePair.Value.SwitchBaseStationIfNeeded();

        IsEnabled = false;
    }

    public virtual void RegisterPhone(Phone phone)
    {
        if (FindPhone(phone.Number) is not null)
            throw new Exception("This number is already connected");

        if (!ConnectedPhonesMap.TryAdd(phone.Number, phone))
            throw new Exception("This number is already connected");

        OnPropertyChanged(nameof(ConnectedPhonesCount));
        OnPropertyChanged(nameof(ConnectedPhones));

        phone.CallEnded += CallEnded;
    }

    public virtual void UnregisterPhone(Phone phone)
    {
        ConnectedPhonesMap.Remove(phone.Number, out _);

        OnPropertyChanged(nameof(ConnectedPhonesCount));
        OnPropertyChanged(nameof(ConnectedPhones));

        phone.CallEnded -= CallEnded;
    }

    public virtual async Task<CallResult> InitiateCall(Phone caller, string to)
    {
        Phone? receiverPhone = FindPhone(to);

        if (receiverPhone is null)
            return CallResult.NotFound;

        if (receiverPhone.CallState is not CallState.Idle)
            return CallResult.Busy;

        CallPhoneData callPhoneData = new(caller, receiverPhone);

        if (!CallDataLookupDictionary.TryAdd(to, callPhoneData))
            return CallResult.Busy;

        bool callerAdded = CallDataLookupDictionary.TryAdd(caller.Number, callPhoneData);

        Debug.Assert(callerAdded);

        CallAnswer callAnswer;

        try
        {
            callAnswer = await receiverPhone.ReceiveCall(caller.Number);
        }
        catch (OperationCanceledException)
        {
            CallDataLookupDictionary.TryRemove(caller.Number, out _);
            CallDataLookupDictionary.TryRemove(to, out _);
            return CallResult.Busy;
        }

        switch (callAnswer)
        {
            case CallAnswer.Agreed:
            {
                return CallResult.Established;
            }
            case CallAnswer.Declined:
            {
                CallDataLookupDictionary.TryRemove(caller.Number, out _);
                CallDataLookupDictionary.TryRemove(to, out _);
                return CallResult.Busy;
            }
            default:
            {
                CallDataLookupDictionary.TryRemove(caller.Number, out _);
                CallDataLookupDictionary.TryRemove(to, out _);
                throw new ArgumentOutOfRangeException(null, "Call answer was out of range");
            }
        }
    }

    protected void CallEnded(object? sender, CallEventArgs e)
    {
        if (!CallDataLookupDictionary.Remove(e.CallerNumber, out CallPhoneData? callData))
            return;

        Debug.Assert(callData is not null);

        bool receiverRemoved = CallDataLookupDictionary.Remove(e.ReceiverNumber, out _);

        Debug.Assert(receiverRemoved);

        callData!.Caller.EndCall();
        callData.Receiver.EndCall();
    }

    protected static Phone? FindPhone(string number)
    {
        Phone? phone = null;
        foreach (BaseStation baseStation in AllEnabledStations)
        {
            if (baseStation.ConnectedPhonesMap.TryGetValue(number, out phone))
                break;
        }

        return phone;
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        // Workaround: WinForms' lack of threading awareness and the challenge of re-invoking handlers in C++/CLI (i.e. no support for lambda-functions)
        if (UiSynchronizationContext != null)
            UiSynchronizationContext.Post(_ => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName)), null);
        else
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    protected bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
            return false;
        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }
}

File: src/TelecommsSimulation.Core/BaseStation.cpp
#include "pch.h"

#include "BaseStation.h"

#include "InvalidBaseStationStateException.h"

using namespace System;
using namespace System::Collections::Generic;
using namespace System::ComponentModel;
using namespace System::Diagnostics;
using namespace System::Linq;
using namespace System::Threading;
using namespace System::Threading::Tasks;

using namespace TelecommsSimulation::Utilities;

namespace TelecommsSimulation::Core
{
    IEnumerable<Phone^>^ BaseStation::ConnectedPhones::get()
    {
        return ConnectedPhonesMap->Values;
    }

    IEnumerable<CallPhoneData^>^ BaseStation::CurrentCalls::get()
    {
        return Enumerable::Distinct<CallPhoneData^>(CallDataLookupDictionary->Values);
    }

    String^ BaseStation::Name::get()
    {
        return _name;
    }

    void BaseStation::Name::set(String^ value)
    {
        SetField<String^>(_name, value, "Name");
    }

    bool BaseStation::IsEnabled::get()
    {
        return _isEnabled;
    }

    void BaseStation::IsEnabled::set(bool value)
    {
        SetField<bool>(_isEnabled, value, "IsEnabled");
    }

    Coordinates BaseStation::Location::get()
    {
        return _location;
    }

    void BaseStation::Location::set(Coordinates value)
    {
        SetField<Coordinates>(_location, value, "Location");
    }

    double BaseStation::CoverageRadiusKm::get()
    {
        return _coverageRadiusKm;
    }

    void BaseStation::CoverageRadiusKm::set(double value)
    {
        SetField<double>(_coverageRadiusKm, value, "CoverageRadiusKm");
    }

    int BaseStation::ConnectedPhonesCount::get()
    {
        return ConnectedPhonesMap->Count;
    }

    IEnumerable<BaseStation^>^ BaseStation::AllEnabledStations::get()
    {
        return AllEnabledStationsList;
    }

    BaseStation::BaseStation(String^ name, Coordinates location, double coverageRadiusKm)
    {
        UiSynchronizationContext = SynchronizationContext::Current;

        Name = name;
        Location = location;
        CoverageRadiusKm = coverageRadiusKm;
    }

    void BaseStation::Enable()
    {
        if (IsEnabled)
            throw gcnew InvalidBaseStationStateException();

        IsEnabled = true;
        AllEnabledStationsList->Add(this);
    }

    void BaseStation::Disable()
    {
        if (!IsEnabled)
            throw gcnew InvalidBaseStationStateException();

        AllEnabledStationsList->Remove(this);

        for each (KeyValuePair<String^, Phone^>^ connectedPhonePair in ConnectedPhonesMap)
            connectedPhonePair->Value->SwitchBaseStationIfNeeded();

        IsEnabled = false;
    }

    void BaseStation::RegisterPhone(Phone^ phone)
    {
        if (FindPhone(phone->Number) != nullptr)
            throw gcnew Exception("This number is already connected");

        if (!ConnectedPhonesMap->TryAdd(phone->Number, phone))
            throw gcnew Exception("This number is already connected");

        OnPropertyChanged("ConnectedPhonesCount");
        OnPropertyChanged("ConnectedPhones");

        phone->CallEnded += gcnew EventHandler<CallEventArgs^>(this, &BaseStation::CallEnded);
    }

    void BaseStation::UnregisterPhone(Phone^ phone)
    {
        Phone^ discard;
        CollectionExtensions::Remove(ConnectedPhonesMap, phone->Number, discard);

        OnPropertyChanged("ConnectedPhonesCount");
        OnPropertyChanged("ConnectedPhones");

        phone->CallEnded -= gcnew EventHandler<CallEventArgs^>(this, &BaseStation::CallEnded);
    }

    Task<CallResult>^ BaseStation::InitiateCall(Phone^ caller, String^ to)
    {
        Phone^ receiverPhone = FindPhone(to);
        if (receiverPhone == nullptr)
            return Task::FromResult(CallResult::NotFound);

        if (receiverPhone->CallState != CallState::Idle)
            return Task::FromResult(CallResult::Busy);

        CallPhoneData^ callPhoneData = gcnew CallPhoneData(caller, receiverPhone);
        if (!CallDataLookupDictionary->TryAdd(to, callPhoneData))
            return Task::FromResult(CallResult::Busy);

        bool callerAdded = CallDataLookupDictionary->TryAdd(caller->Number, callPhoneData);
        Debug::Assert(callerAdded);

        Task<CallAnswer>^ receiveCallTask = receiverPhone->ReceiveCall(caller->Number);
        return receiveCallTask->ContinueWith<CallResult>(
            gcnew Func<Task<CallAnswer>^, Object^, CallResult>(this, &BaseStation::InitiateCallContinuation),
            CallInitiationContinuationContext(to, caller, callPhoneData),
            TaskContinuationOptions::RunContinuationsAsynchronously);
    }

    CallResult BaseStation::InitiateCallContinuation(Task<CallAnswer>^ continuedTask, Object^ continuationContext)
    {
        CallInitiationContinuationContext context = safe_cast<CallInitiationContinuationContext>(continuationContext);

        CallPhoneData^ discard;

        if (continuedTask->IsCanceled || continuedTask->IsFaulted)
        {
            CallDataLookupDictionary->TryRemove(context.Caller->Number, discard);
            CallDataLookupDictionary->TryRemove(context.To, discard);
            return CallResult::Busy;
        }

        switch (continuedTask->Result)
        {
            case CallAnswer::Agreed:
                {
                    return CallResult::Established;
                }
            case CallAnswer::Declined:
                {
                    CallDataLookupDictionary->TryRemove(context.Caller->Number, discard);
                    CallDataLookupDictionary->TryRemove(context.To, discard);
                    return CallResult::Busy;
                }
            default:
                {
                    CallDataLookupDictionary->TryRemove(context.Caller->Number, discard);
                    CallDataLookupDictionary->TryRemove(context.To, discard);
                    throw gcnew ArgumentOutOfRangeException(nullptr, "Call answer was out of range");
                }
        }
    }

    void BaseStation::CallEnded(Object^ sender, CallEventArgs^ e)
    {
        CallPhoneData^ callData;

        if (!CollectionExtensions::Remove(CallDataLookupDictionary, e->CallerNumber, callData))
            return;

        Debug::Assert(callData != nullptr);

        const bool receiverRemoved =
            CollectionExtensions::Remove(CallDataLookupDictionary, e->ReceiverNumber, callData);

        Debug::Assert(receiverRemoved);

        callData->Caller->EndCall();
        callData->Receiver->EndCall();
    }

    Phone^ BaseStation::FindPhone(String^ number)
    {
        Phone^ phone = nullptr;
        for each (BaseStation^ baseStation in AllEnabledStations)
        {
            if (baseStation->ConnectedPhonesMap->TryGetValue(number, phone))
                break;
        }

        return phone;
    }

    void BaseStation::OnPropertyChanged(String^ propertyName)
    {
        // Workaround: WinForms' lack of threading awareness and the challenge of re-invoking handlers in C++/CLI
        if (UiSynchronizationContext != nullptr)
        {
            UiSynchronizationContext->Post(
                gcnew SendOrPostCallback(this, &BaseStation::OnPropertyChangedInvoke),
                propertyName);
        }
        else
        {
            PropertyChanged(this, gcnew PropertyChangedEventArgs(propertyName));
        }
    }

    void BaseStation::OnPropertyChangedInvoke(Object^ propertyName)
    {
        PropertyChanged(this, gcnew PropertyChangedEventArgs(safe_cast<String^>(propertyName)));
    }

    generic<typename T>
    bool BaseStation::SetField(T% field, T value, String^ propertyName)
    {
        if (EqualityComparer<T>::Default->Equals(field, value))
            return false;

        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }
}
